File: frontend/src/app/models/tour.ts
export type TransportType =
  | 'DRIVING_CAR'
  | 'CYCLING_REGULAR'
  | 'FOOT_WALKING'
  | 'FOOT_HIKING';

export interface RouteGeometry {
  type: string;
  coordinates: number[][];   // [[lon,lat], ...]
}

export interface Tour {
  id: number;
  name: string;
  description: string;
  fromName: string;
  fromLat: number;
  fromLon: number;
  toName: string;
  toLat: number;
  toLon: number;
  transportType: TransportType;
  distanceKm: number;
  durationSec: number;
  routeGeoJson: RouteGeometry | null;
  imageUrl: string | null;
  ownerId: number;
  popularity: number;
  childFriendly: boolean | null;
}

export interface CreateTourRequest {
  name: string;
  description: string;
  fromName: string;
  fromLat: number;
  fromLon: number;
  toName: string;
  toLat: number;
  toLon: number;
  transportType: TransportType;
}

File: backend/src/main/java/fhtw/swen2/model/Tour.java
package fhtw.swen2.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import com.fasterxml.jackson.annotation.JsonManagedReference;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "tours")
@Getter
@Setter
@NoArgsConstructor
public class Tour {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @NotBlank
    @Column(nullable = false)
    private String name;

    @NotBlank
    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @NotBlank
    @Column(nullable = false)
    private String fromName;

    @Column(nullable = false)
    private double fromLat;

    @Column(nullable = false)
    private double fromLon;

    @NotBlank
    @Column(nullable = false)
    private String toName;

    @Column(nullable = false)
    private double toLat;

    @Column(nullable = false)
    private double toLon;


    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransportType transportType;

    @Column(nullable = false)
    private double distanceKm;

    @Column(nullable = false)
    private long durationSec;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private RouteGeometry routeGeoJson;

    private String imageFilename;

    @ManyToOne(optional = false)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @JsonManagedReference
    @OneToMany(mappedBy = "tour", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TourLog> logs = new ArrayList<>();
}
